kotlinx-coroutines-core/common/src/CoroutineStart.kt

```diff
@@ -1,74 +1,342 @@
+     * never be completed, leading to deadlocks and resource leaks.
+     * For example, the following code will deadlock, since [coroutineScope] waits for all of its child coroutines to
+     * complete:
+     * ```
+     * coroutineScope {
+     *     launch(start = CoroutineStart.LAZY) { }
+     * }
+     * ```
+     *
+     * Behavior of [LAZY] can be described with the following examples:
+     *
+     * ```
+     * // Example of lazily starting a new coroutine that goes through a dispatch
+     * runBlocking {
+     *     println("1. About to start a new coroutine.")
+     *     // Create a job to execute on `Dispatchers.Default` later.
+     *     val job = launch(Dispatchers.Default, start = CoroutineStart.LAZY) {
+     *         println("3. Only now does the coroutine start.")
+     *     }
+     *     delay(10.milliseconds) // try to give the coroutine some time to run
+     *     println("2. The coroutine still has not started. Now, we await it.")
+     *     job.join()
+     * }
+     * ```
+     *
+     * ```
+     * // Example of lazily starting a new coroutine that doesn't go through a dispatch initially
+     * runBlocking {
+     *     println("1. About to lazily start a new coroutine.")
+     *     // Create a job to execute on `Dispatchers.Unconfined` later.
+     *     val lazyJob = launch(Dispatchers.Unconfined, start = CoroutineStart.LAZY) {
+     *         println("3. The coroutine starts on the thread that called `join`.")
+     *     }
+     *     // We start the job on another thread for illustrative purposes
+     *     launch(Dispatchers.Default) {
+     *         println("2. We start the lazyJob.")
+     *         job.start() // runs lazyJob's code in-place
+     *         println("4. Only now does the `start` call return.")
+     *     }
+     * }
+     * ```
      */
     LAZY,
 
     /**
      * Atomically (i.e., in a non-cancellable way) schedules the coroutine for execution according to its context.
-     * This is similar to [DEFAULT], but the coroutine cannot be cancelled before it starts executing.
      *
-     * Cancellability of coroutine at suspension points depends on the particular implementation details of
-     * suspending functions as in [DEFAULT].
+     * This is similar to [DEFAULT], but the coroutine is guaranteed to start executing even if it was cancelled.
+     * This only affects the initial portion of the code: on subsequent suspensions, cancellation will work as usual.
+     *
+     * [UNDISPATCHED] also ensures that coroutines will be started in any case.
+     * The difference is that, instead of immediately starting them on the same thread,
+     * [ATOMIC] performs the full dispatch procedure just as [DEFAULT] does.
+     *
+     * Because of this, we can use [ATOMIC] in cases where we want to be certain that some code eventually runs
+     * and uses a specific dispatcher to do that.
```

Comment:
This sample is a little bit problematic because it contains non-trivial logic that obscures the actual atomic logic.

I would suggest the example from the `DEFAULT` but with something like `This code will execute even though the coroutine was cancelled before the dispatch` instead of `This code will never execute`.

Also, let's leave a warning comment (without an example) that it might be bug-prone if used e.g. with UI elements that are destroyed as part of the cancellatiobn

Reply:
>but I don't understand the difference between this sample and something like `currentSystemDefault`

Datetime example contains two lines that are relevant to the specific use-case (`return "[$formattedTime] $message"` and `logEntry("Starting the application")`), and they are quite easy to skim through. Everything else is focused solely on the relevant datetime concepts.

Here it is different -- a semaphore (that is, arguably, a more advanced concept and definitely much less popular concept), non-trivial branchy logic, two `finally` blocks, `NonCancellable` etc. Understanding both the sample and why it is written in such a way is harder than the typical `ATOMIC` and also expects you to both know the other parts of coroutines and how to use some stuff correctly. From our experience, an unprepared reader is likely to skip is completely -- the sample has to be really straightforward.

Reply:
I see. Do you think we should try cooking up a more straightforward sample or just abandon the idea of highlighting the use case here?

Reply:
I would suggest giving a straightforward example without a business logic. You have a perfect one for `DEFAULT`, `ATOMIC` might use quite a similar one, but where cancellation has no effect during the dispatching

Reply:
It might still be worth highlighting this "using `ATOMIC` to clean up resources acquired outside of a coroutine" pattern, though. I think a good compromise would be to collapse this sample to something like this:

```kotlin
@OptIn(ExperimentalCoroutinesApi::class)
fun main() {
    runBlocking {
        val mutex = Mutex()
        mutex.lock()
        val job = launch(start = CoroutineStart.ATOMIC) { // the coroutine is scheduled for execution
            println("Coroutine running!")
            try {
                // this `try` block will be entered in any case because of ATOMIC
                println("Starting task...")
                delay(10.milliseconds) // throws due to cancellation
                println("Finished task.")
            } finally {
                mutex.unlock() // returns the lock that was aquired before starting the coroutine correctly
            }
        }
        job.cancelAndJoin() // we immediately cancel the coroutine.
        mutex.withLock {
            println("The lock has been returned correctly!")
        }
    }
}
```

Reply:
Changes looko good (and explanations are definitely improved) – any reason why this example isn't wrapped in a `runBlocking`?

Reply:
The idea was to use `runBlocking` in the samples where we want to show deterministic behavior, focusing on the interplay between `CoroutineStart` and dispatchers, but here, we are highlighting a realistic scenario that could well be parallel. If we add `runBlocking` here, a natural question may arise: is this code only correct because it's running on one thread?

Reply:
Fair enough!